### PR TITLE
`@mem{cpy,set}` fixes

### PR DESCRIPTION
**Pull request by Vexu**
Closes #15638, with parts being reused with attribution

### CODE REVIEW

File: src/codegen/llvm.zig [hunk 1]
```diff
@@ -8588,7 +8588,10 @@ pub const FuncGen = struct {
         const src_slice = try self.resolveInst(bin_op.rhs);
         const src_ptr_ty = self.air.typeOf(bin_op.rhs);
         const src_ptr = self.sliceOrArrayPtr(src_slice, src_ptr_ty);
-        const len = self.sliceOrArrayLenInBytes(dest_slice, dest_ptr_ty);
```
Comment (andrewrk) [inline src/codegen/llvm.zig:8591]: AIR docs for memcpy say:
> Dest pointer is either a slice or a pointer to array.

File: test/cases/compile_errors/incorrect_type_to_memset_memcpy.zig [hunk 2]
```diff
@@ -27,10 +40,14 @@ pub export fn non_matching_lengths() void {
 // :5:5: error: unknown @memcpy length
 // :5:18: note: destination type '[*]u8' provides no length
 // :5:24: note: source type '[*]align(4) const u8' provides no length
-// :10:13: error: type 'u8' does not support indexing
-// :10:13: note: operand must be an array, slice, tuple, or vector
-// :15:13: error: type '*u8' does not support indexing
-// :15:13: note: operand must be an array, slice, tuple, or vector
+// :10:13: error: type '*u8' is not an indexable pointer
+// :10:13: note: operand must be a slice, a many pointer or a pointer to an array
```
Comment (andrewrk) [inline test/cases/compile_errors/incorrect_type_to_memset_memcpy.zig:44]: Destination may be a tuple, previous message was more correct

Comment (Vexu) [reply test/cases/compile_errors/incorrect_type_to_memset_memcpy.zig:44]: My understanding was that that hasn't been implemented yet #15479?

File: src/codegen/llvm.zig (hunk 1 above)
Comment (Vexu) [reply src/codegen/llvm.zig:8591]: The language reference says that a many pointer is allowed https://ziglang.org/documentation/master/#memcpy, should that be changed?

Comment (andrewrk) [reply src/codegen/llvm.zig:8591]: no - a many pointer is allowed in the language but semantic analysis will ensure the length property will be communicated always by the dest argument.

File: test/cases/compile_errors/incorrect_type_to_memset_memcpy.zig (hunk 2 above)
Comment (andrewrk) [reply test/cases/compile_errors/incorrect_type_to_memset_memcpy.zig:44]: That's correct, it's not implemented yet. It's fine to do as you have it, as long as you're aware :+1:

Comment (Vexu) [reply test/cases/compile_errors/incorrect_type_to_memset_memcpy.zig:44]: There is a link to the issue next to the error message as a reminder.

**Pull request merged**